lib/recursive-methods.rb

```diff
@@ -1,49 +1,84 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n), where n is the number handed to the method.
+# Space complexity: O(n) as well
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+    raise ArgumentError if n < 0
+    return 1 if n == 0 || n == 1
+    return n * factorial(n-1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n^2), where n is the length of the string passed
+# Space complexity: O(n^2)
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+    return s if s.length <= 1
+    return s[-1] + reverse(s[0..-2])
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n^2), where n is the length of the string passed .. does the replace impact this as well??
+# Space complexity: O(n^2)
 def reverse_inplace(s)
-    raise NotImplementedError, "Method not implemented"
+    return s if s.length <= 1
+    n = s[-1] + reverse_inplace(s[0..-2])
+    return s.replace n
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) where n is the number passed to the method
+# Space complexity: O(n) as well
 def bunny(n)
-    raise NotImplementedError, "Method not implemented"
+    return 0 if n == 0
+    return 2 if n == 1
+    return 2 + bunny(n-1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n^2) where n is the length of the string
+# Space complexity: O(n^2)
 def nested(s)
```

Comment:
👍 However can you see a way to do this with O(n) time/space complexity similar to how you can do reverse_inplace?